### DISCUSSION TITLE
Questions on LibIGL

### DISCUSSION THREAD

**Discussion by drathin4** [libigl/libigl#1812] [Q&A]
I have 3 questions related to LibIGL:

1. From the perspective of mesh boolean, if input mesh is not `PWN` (piecewise constant winding number), can we use any mesh repairing APIs to make it `PWN` ? I have tried using the APIs `is_irregular_vertex(), is_vertex_manifold(), is_vertex_manifold()`, to determine any issues in the mesh, but none give any problems, yet mesh boolean algorithm asserts that it is `PWN`.
2. How to set appropriate tolerance for `remove_duplicate_vertices()` ? What tolerance does LibIGL use ? I have been using a tolerance of 1e-5 and have been getting reasonable results, but I want to know if there's any rigorous mathematical technique to determine tolerance.
3. Is it possible to query mesh colors and if yes, how ? I'm looking for the `getter` equivalent of `set_colors()`.

**Comment by alecjacobson** [comment]
1. libigl doesn't have a generic "auto-fix" for this. I tend to use https://github.com/MarcoAttene/MeshFix-V2.1, though sometimes it is too aggressive. If you like to diagnose your problems you might look at the [`meshstatistics` tool of gp-cli](https://github.com/alecjacobson/gp-cli#meshstatistics). Often the issue is at open boundaries or conflictingly oriented/non-manifold edges.

2. The tolerance of `remove_duplicate_vertices` is in absolute units so it will depend on the scale of your input and how separated your duplicates are. It will also matter how you ended up with duplicates. For example `.stl` files have no concept of shared vertices so when reading in their triangle soup, using a tolerance of `0` will be fine assuming that the written .stl started as a valid mesh. For efficiency reasons, `remove_duplicate_vertices` considers duplicity per-coordinate rather than L2 distance.

3. The internal variables are public so you can access them: [`*_material_*`](https://github.com/libigl/libigl/blob/main/include/igl/opengl/ViewerData.h#L206) depending on whether the mesh is using face- or vertex-based colors and whether you want ambient, diffuse, specular.
